desci-server/src/controllers/nodes/createDpid.ts

```diff
@@ -0,0 +1,159 @@
+import { Response } from "express";
+import { ethers } from "ethers";
+import { logger as parentLogger } from '../../logger.js';
+import { RequestWithNode } from "../../middleware/authorisation.js";
+import { contracts, typechain as tc } from "@desci-labs/desci-contracts";
+import { DpidMintedEvent, UpgradedDpidEvent } from "@desci-labs/desci-contracts/dist/typechain-types/DpidAliasRegistry.js";
+import { setDpidAlias } from "../../services/nodeManager.js";
+
+type DpidResponse = DpidSuccessResponse | DpidErrorResponse;
+export type DpidSuccessResponse = {
+  dpid: number;
+};
+
+export type DpidErrorResponse = {
+  error: string;
+};
+
+/** Not secret: pre-seeded ganache account for local dev */
+const GANACHE_PKEY = "ac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
+
+let aliasRegistryAddress: string;
+const url = process.env.SERVER_URL;
+
+if (url.includes("localhost")) {
+  aliasRegistryAddress = contracts.localDpidAliasInfo.proxies.at(0).address;
+} else if (url.includes("dev") || url.includes("staging")) {
+  aliasRegistryAddress = contracts.devDpidAliasInfo.proxies.at(0).address;
+} else if (process.env.NODE_ENV === "production") {
+  aliasRegistryAddress = contracts.prodDpidAliasInfo.proxies.at(0).address;
+};
+
+export const createDpid = async (req: RequestWithNode, res: Response<DpidResponse>) => {
+  const owner = req.user;
+  const node = req.node;
+  const { uuid } = req.body;
+
+  const logger = parentLogger.child({
+    module: "NODE::createDpidController",
+    body: req.body,
+    uuid,
+    user: owner,
+    ceramicStream: node.ceramicStream,
+  });
+
+  if (!uuid) {
+    return res.status(400).json({ error: "UUID is required" });
+  };
+
+  if (!process.env.HOT_WALLET_KEY) {
+    logger.error("hot wallet not configured");
+    return res.status(500).json({ error: "dpid registration not available" });
+  };
+
+  if (!process.env.ETHEREUM_RPC_URL) {
+    logger.error("ethereum RPC endpoint not configured");
+    return res.status(500).json({ error: "dpid registration not available" });
+  };
+
+  try {
+    const dpid = await getOrCreateDpid(node.ceramicStream);
+    if (!node.dpidAlias) {
+      setDpidAlias(uuid, dpid);
+    };
+    return res.status(200).send({ dpid });
+  } catch (err) {
+    logger.error({ err }, "node-create-dpid-err");
+    return res.status(400).send({ error: err.message });
+  };
+};
+
+export const getOrCreateDpid = async (
+  streamId: string,
+): Promise<number> => {
+  const logger = parentLogger.child({
+    module: "NODE::mintDpid",
+    ceramicStream: streamId,
+  });
+
+  const provider = new ethers.providers.JsonRpcProvider(
+    process.env.ETHEREUM_RPC_URL
+  );
+
+  await provider.ready;
+  const wallet = new ethers.Wallet(
+    url.includes("localhost") ? GANACHE_PKEY : process.env.HOT_WALLET_KEY,
```

Comment:
we do this logic in multiple places, consider moving it to the top and having a var we can reuse